File: leetcode-cn/Python/146.lru缓存机制.py
#
# @lc app=leetcode.cn id=146 lang=python3
#
# [146] LRU缓存机制
#

# @lc code=start


class LinkedNode:
    def __init__(self, key=0, value=0):
        self.key = key
        self.value = value
        self.pre = None
        self.next = None


class LRUCache:
    def __init__(self, capacity: int):
        self.head = LinkedNode()
        self.tail = LinkedNode()

        self.head.next = self.tail
        self.tail.pre = self.head

        self.capacity = capacity
        self.size = 0
        self.cache = {}

    def get(self, key: int) -> int:
        if key in self.cache:
            node = self.cache[key]
            self.move2head(node)
            return node.value
        else:
            return -1

    def put(self, key: int, value: int) -> None:
        if key not in self.cache:
            node = LinkedNode(key, value)
            self.cache[key] = node
            self.add_head(node)

            self.size += 1

            if self.size > self.capacity:
                removed = self.remove_tail()
                self.cache.pop(removed.key)
                self.size -= 1

        else:
            node = self.cache[key]
            node.value = value
            self.move2head(node)

    def add_head(self, node):
        node.pre = self.head
        node.next = self.head.next
        self.head.next.pre = node
        self.head.next = node

    def remove_node(self, node):
        node.pre.next = node.next
        node.next.pre = node.pre

    def move2head(self, node):
        self.remove_node(node)
        self.add_head(node)

    def remove_tail(self):
        node = self.tail.pre
        self.remove_node(node)

        return node


# Your LRUCache object will be instantiated and called as such:
# obj = LRUCache(capacity)
# param_1 = obj.get(key)
# obj.put(key,value)
# @lc code=end

File: leetcode-cn/C++/146.lru-缓存机制.cpp
/*
 * @lc app=leetcode.cn id=146 lang=cpp
 *
 * [146] LRU 缓存机制
 */

// @lc code=start
#include <unordered_map>
using namespace std;

struct DListNode
{
    DListNode *pre;
    DListNode *next;
    int val;
    int key;
};

class LRUCache
{
public:
    unordered_map<int, DListNode *> mem;
    int capacity = 0;
    int size = 0;
    DListNode *head;
    DListNode *tail;

    LRUCache(int capacity)
    {
        this->capacity = capacity;

        head = new DListNode();
        tail = new DListNode();
        head->next = tail;
        tail->pre = head;
    }

    void deleteNode(DListNode *node)
    {
        node->pre->next = node->next;
        node->next->pre = node->pre;

        node->pre = nullptr;
        node->next = nullptr;
    }

    void move2head(DListNode *node)
    {
        node->next = head->next;
        head->next->pre = node;

        head->next = node;
        node->pre = head;
    }

    int get(int key)
    {
        if (mem.count(key))
        {
            DListNode *node = mem[key];
            deleteNode(node);
            move2head(node);
            return node->val;
        }
        else
        {
            return -1;
        }
    }

    void put(int key, int value)
    {
        if (mem.count(key))
        {
            DListNode *node = mem[key];
            node->val = value;
            deleteNode(node);
            move2head(node);
        }
        else
        {
            if (size == capacity)
            {
                DListNode *node = tail->pre;
                deleteNode(node);
                mem.erase(node->key);
                delete node;
                size--;
            }

            DListNode *new_node = new DListNode();
            new_node->val = value;
            new_node->key = key;
            mem[key] = new_node;

            move2head(new_node);
            size++;
        }
    }
};

/**
 * Your LRUCache object will be instantiated and called as such:
 * LRUCache* obj = new LRUCache(capacity);
 * int param_1 = obj->get(key);
 * obj->put(key,value);
 */
// @lc code=end
